tt_metal/host_api.hpp

```diff
@@ -189,14 +189,16 @@ uint32_t CreateSemaphore(Program &program, const std::variant<CoreRange,CoreRang
 *
 *  Return value: Buffer
 *
-*  | Argument    | Description                             | Type       | Valid Range | Required |
-*  |-------------|---------------------------------------- |------------|-------------|----------|
-*  | device      | The device that the buffer will reside  | Device     |             | Yes      |
-*  | size        | size of buffer                          | uint64_t   |             | Yes      |
-*  | page_size   | buffer page size                        | uint64_t   |             | Yes      |
-*  | buffer_type | type of buffer (L1 or DRAM)             | BufferType |             | Yes      |
+*  | Argument        | Description                             | Type                     | Valid Range | Required |
+*  |-----------------|---------------------------------------- |--------------------------|-------------|----------|
+*  | device          | The device that the buffer will reside  | Device                   |             | Yes      |
+*  | size            | size of buffer                          | uint64_t                 |             | Yes      |
+*  | page_size       | buffer page size                        | uint64_t                 |             | Yes      |
+*  | buffer_type     | type of buffer (L1 or DRAM)             | BufferType               |             | Yes      |
+*  | buffer_layout   | layout, default is INTERLEAVED          | TensorMemoryLayout       |             | No       |
+*  | shard_parameter | parameters required for sharding        | std::optional<ShardSpec> |             | No       |
 */
-Buffer CreateBuffer(Device *device, std::uint64_t size, std::uint64_t page_size, const BufferType buffer_type);
+Buffer CreateBuffer(Device *device, std::uint64_t size, std::uint64_t page_size, const BufferType buffer_type, const TensorMemoryLayout buffer_layout = TensorMemoryLayout::INTERLEAVED, std::optional<ShardSpecBuffer> shard_parameter= std::nullopt);
```

Comment:
to keep in line with the other CreateX APIs I wonder if we should add BufferConfig and then CreateBuffer accepts Device and BufferConfig. 

Thoughts @TT-billteng

Reply:
like this idea, do you see other possible configs in the future? @tarafdarTT @abhullar-tt

Reply:
I think it would be within BufferConfig, but we might have different types of buffers. Currently for sharding across L1 we have specific types of mappings, but for multi-device mappings our mapping structs could change, but all of that is within the BufferConfig

Reply:
do we see any additional config parameters for standard DRAM and/or L1 buffers (not sharded)? Right now I see possible to have:

`DRAMConfig` (empty struct)
`L1Config` (empty struct)
`L1Sharded` 
??

Reply:
or you think we should have `Device` be a member of all these configs @abhullar-tt ?

Reply:
in the short term no. but in the long term with multi-device we might want to have shards across multiple DRAM address spaces.

Reply:
maybe we can have `InterleavedConfig` and `ShardedConfig`. InterleavedConfig can accept whether buffer is in dram or l1. Sharded for now can mean L1 and eventually if we want DRAM shards we can add that in there as well. 

would be a good idea to just add device into the buffer configs right now, even though we only ever allocate  buffer on one device. that way for multi-device sharded buffer there won't be an API update to `CreateBuffer`, only the config would need to be updated

Reply:
this is quite a big change (in 87 files I'll need to change how it calls the API). I've changed about 20 or so of the files so far. Before I continue just want to confirm from you guys @abhullar-tt @TT-billteng 

```
                tt_metal::BufferConfig l1_config{
                                        .device=device,
                                        .size = per_core_l1_size,
                                        .page_size = per_core_l1_size,
                                        .buffer_type = tt_metal::BufferType::L1 
                                        };
                auto l1_b0 = CreateBuffer(l1_config);
```

Reply:
and ShardedBufferConfig would extent BufferConfig. with default BufferConfig specifying Interleaved buffers?

Reply:
oh I was gonna have ShardSpec as an optional inside buffer config, but I can do your suggestion as well.

Reply:
how about this?


API: 

```
Buffer CreateBuffer(std::variant<InterleavedBufferConfig, ShardedBufferConfig> config);
```

Config Structs:
```
struct BufferConfig {
    Device *device;
    uint64_t size;                 // Size in bytes
    uint64_t page_size;            // Size of unit being interleaved. For non-interleaved buffers: size == page_size
    BufferType buffer_type;
    TensorMemoryLayout buffer_layout = TensorMemoryLayout::INTERLEAVED;
};

typedef BufferConfig InterleavedBufferConfig;

struct ShardedBufferConfig : BufferConfig {
    ShardSpecBuffer shard_parameters;
};
```

Reply:
good with me!

Reply:
looks good :)